### PR TITLE
[mono][wasm] Add beginnings of support for WASI.

### PR DESCRIPTION
**Pull request by vargaz**

### CODE REVIEW

File: src/mono/CMakeLists.txt [hunk 1]
```diff
@@ -231,6 +231,16 @@ elseif(CMAKE_SYSTEM_NAME STREQUAL "Emscripten")
   # sys/random.h exists, but its not found
   set(HAVE_SYS_RANDOM_H 1)
   set(INTERNAL_ZLIB 1)
+elseif(CMAKE_SYSTEM_NAME STREQUAL "WASI")
+  set(HOST_WASI 1)
+  add_definitions(-D_WASI_EMULATED_SIGNAL -D_WASI_EMULATED_MMAN)
+  add_definitions(-DNO_GLOBALIZATION_SHIM)
+  add_definitions(-D_THREAD_SAFE)
+  # FIXME: Is there a cmake option for this ?
+  set(DISABLE_SHARED_LIBS 1)
```
Comment (am11) [inline src/mono/CMakeLists.txt:240]: From https://cmake.org/cmake/help/latest/guide/tutorial/Selecting%20Static%20or%20Shared%20Libraries.html
```suggestion
  option(BUILD_SHARED_LIBS "Disabling shared libraries for WASI" OFF)
```

**Comment by pavelsavara** [comment]
Would it be different RID we could make it compatible with `/p:TargetOS=Browser /p:TargetArchitecture=wasm` ?

**Comment by vargaz** [comment]
It's a very different environment, there is no JS, and currently, a lot of things are missing, like networking, threads etc.

**Comment by vargaz** [comment]
With these changes, the build still fails because a lot of things are missing like pthread.h, socket functions, etc.

**Review by lambdageek — APPROVED** [review]

**Pull request merged**